samples/boards/stm32/bluetooth/interactive_gui/src/main.c

```diff
@@ -0,0 +1,473 @@
+				header_length = hdr_len(type);
+				if (type == H4_ST_EXT_CMD) {
+					/* Convert to regular HCI_CMD */
+					if (remaining > 255) {
+						LOG_ERR("len > 255");
+						net_buf_unref(buf);
+						state = ST_DISCARD;
+					} else {
+						header_length--;
+					}
+				}
+				net_buf_add_mem(buf, hdr_buf, header_length);
+				if (remaining > net_buf_tailroom(buf)) {
+					LOG_ERR("Not enough space in buffer");
+					net_buf_unref(buf);
+					state = ST_DISCARD;
+				} else {
+					state = ST_PAYLOAD;
+				}
+
+			}
+			break;
+		case ST_PAYLOAD:
+			read = h4_read(hci_uart_dev, net_buf_tail(buf), remaining);
+			buf->len += read;
+			remaining -= read;
+			if (remaining == 0) {
+				/* Packet received */
+				LOG_DBG("putting RX packet in queue.");
+				net_buf_put(&tx_queue, buf);
+				state = ST_IDLE;
+			}
+			break;
+		case ST_DISCARD:
+			uint8_t discard[H4_DISCARD_LEN];
+			size_t to_read = MIN(remaining, sizeof(discard));
+
+			read = h4_read(hci_uart_dev, discard, to_read);
+			remaining -= read;
+			if (remaining == 0) {
+				state = ST_IDLE;
+			}
+			break;
+		default:
+			read = 0;
+			__ASSERT_NO_MSG(0);
+			break;
+
+		}
+	} while (read);
+}
+
+static void tx_isr(void)
+{
+	static struct net_buf *buf;
+	int len;
+
+	if (!buf) {
+		buf = net_buf_get(&uart_tx_queue, K_NO_WAIT);
+		if (!buf) {
+			uart_irq_tx_disable(hci_uart_dev);
+			return;
+		}
+	}
+	len = uart_fifo_fill(hci_uart_dev, buf->data, buf->len);
+	net_buf_pull(buf, len);
+	if (!buf->len) {
+		net_buf_unref(buf);
+		buf = NULL;
+	}
+}
+
+static void bt_uart_isr(const struct device *unused, void *user_data)
+{
+	ARG_UNUSED(unused);
+	ARG_UNUSED(user_data);
+
+	if (!(uart_irq_rx_ready(hci_uart_dev) || uart_irq_tx_ready(hci_uart_dev))) {
+		LOG_DBG("spurious interrupt");
+	}
+	if (uart_irq_tx_ready(hci_uart_dev)) {
+		tx_isr();
+	}
+	if (uart_irq_rx_ready(hci_uart_dev)) {
+		rx_isr();
+	}
+}
+
+static void tx_thread(void *p1, void *p2, void *p3)
+{
+	enum bt_buf_type buf_type;
+
+	while (1) {
+		struct net_buf *buf;
+		int err = 0;
+		uint8_t len;
+		uint8_t response[16];
+
+		/* Wait until a buffer is available */
+		buf = net_buf_get(&tx_queue, K_FOREVER);
+		buf_type = bt_buf_get_type(buf);
+		if (buf_type == H4_ST_VND_CMD) {
+			len = parse_cmd(buf->data, buf->len, response);
+			err =  send_evt(response, len);
+			if (!err) {
+				net_buf_unref(buf);
+			}
+		} else {
+			/* Pass buffer to the stack */
+			err = bt_send(buf);
+		}
+		if (err) {
+			LOG_ERR("Unable to send (err %d)", err);
+			net_buf_unref(buf);
+		}
+
+		/* Give other threads a chance to run if tx_queue keeps getting
+		 * new data all the time.
+		 */
+		k_yield();
+	}
+}
+
+static int h4_send(struct net_buf *buf)
+{
+	LOG_DBG("buf %p type %u len %u", buf, bt_buf_get_type(buf), buf->len);
+	net_buf_put(&uart_tx_queue, buf);
+	uart_irq_tx_enable(hci_uart_dev);
+	return 0;
+}
+
+static int hci_uart_init(void)
+{
+	LOG_DBG("");
+	if (IS_ENABLED(CONFIG_USB_CDC_ACM)) {
```

Comment:
I don't see you enabling USB support in the sample's `prj.conf`. Is it implicitly enabled for these boards? If not, it'd be good to have CI coverage of all these conditional branches, so adding another .conf file (or enabling USB in the existing one) might be a good thing to do.

Reply:
I removed it as it was remained from `hci_uart` sample. Instead, I added `CONFIG_USB_DEVICE_INITIALIZE_AT_BOOT=y` to the `prj.conf` file.